src/fastpath/fastpath.c

```diff
@@ -12,6 +12,192 @@
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_EXCEPTION_FASTPATH
+static inline
+FORCE_INLINE
+void NORETURN fastpath_vm_fault(vm_fault_type_t type)
+{
+    cap_t handler_cap;
+    endpoint_t *ep_ptr;
+    tcb_t *dest;
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    word_t badge;
+    seL4_MessageInfo_t info;
+    word_t msgInfo;
+    pde_t stored_hw_asid;
+    dom_t dom;
+
+    /* Get the fault handler endpoint */
+#ifdef CONFIG_KERNEL_MCS
+    handler_cap = TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbFaultHandler)->cap;
+#else
+    cptr_t handlerCPtr;
+    handlerCPtr = NODE_STATE(ksCurThread)->tcbFaultHandler;
+    handler_cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, handlerCPtr);
+#endif
+
+    if (unlikely(!cap_capType_equals(handler_cap, cap_endpoint_cap) ||
+                 !cap_endpoint_cap_get_capCanReceive(handler_cap))) {
```

Comment:
This cap right check needs to do different things for MCS and for non-MCS.  On MCS the cap rights are checked when the cap is added to the TCB, so checking that the type is cap_endpoint_cap is sufficient.
For non-mcs, the checks need to be that the cap is an endpoint cap AND the cap has send rights, AND the cap has (either grant rights or grantreply rights):  
```
        cap_endpoint_cap_get_capCanSend(handlerCap) &&
        (cap_endpoint_cap_get_capCanGrant(handlerCap) ||
         cap_endpoint_cap_get_capCanGrantReply(handlerCap)))
```